File: DataStructures/Queue_Stack/Stack/stack.js
class Stack {
    constructor() {
        this._size = 0;
        this._storage = {};
    }
    /**
     * Push a value onto the end of the stack
     * assigns size as a key of storage
     * assigns value as the value of this key
     * Increase the size of the storage
     */
    push(data) {
        this._storage[this._size] = data;
        this._size++;
    }

    // Remove from the stack - pop()
    pop() {
        if (this._size && this._size !== 0) {
            const result = this._storage[this._size];
            delete this._storage[this._size];
            this._size--;
            return result;
        }
        return null;
    }
}

const stack = new Stack();
// Push
stack.push(1);
stack.push(23);
console.log(stack);

// Pop
stack.pop(23)
console.log(stack);

File: DataStructures/Queue_Stack/Stack/stack.cpp
#include <iostream>

class Stack {
    int *arr;
    int pos;
    int capacity;
    public:
        Stack(int);
        ~Stack();
        void push(int);
        int pop();
        int size();
        int peek();
};

Stack::Stack(int size) {
    arr = new int[size];
    pos = -1;
    capacity = size;
}

Stack::~Stack() {
    delete []arr;
}

void Stack::push(int value) {
    if(pos > capacity) {
        std::cout << "Stack overflow" << std::endl;
        return;
    }
    arr[++pos] = value;
    return;
} 

int Stack::pop() {
    if(pos < 0) {
        std::cout << "Stack is empty" << std::endl;
        return 0;
    }
    return arr[pos--];
}

int Stack::peek() {
    return arr[pos];
}

int Stack::size() {
    return pos + 1;
}


int main() {
    Stack s(3);
    s.push(1);
    s.push(2);
    s.push(3);
    s.pop();
    s.push(4);
    std::cout << "Stack size is: " << s.size() << std::endl;
}